BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -122,26 +198,63 @@ uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
  * @pre     The data parameter must be at least 8 bytes or hardfault may occur.
  * @param   id : pointer to store id of the message that was received.
  * @param   data : pointer to store data that was received. Must be 8bytes or bigger.
- * @return  0 if nothing was received so ignore id and data that was received. Any other value indicates data was received and stored.
+ * @return  ERROR if nothing was received so ignore id and data that was received. SUCCESS indicates data was received and stored.
  */
-uint8_t BSP_CAN_Read(uint32_t *id, uint8_t *data) {
-    if(RxFlag){
-		for(int i = 0; i < 8; i++){
-			data[i] = RxMessage.Data[i];
-		}
-        *id = RxMessage.StdId;
-		RxFlag = false;
-		return 1;
-	}
-    return 0;
+ErrorStatus BSP_CAN_Read(uint32_t *id, uint8_t *data) {
+    // If the queue is empty, return err
+    if(gRxQueue.head == gRxQueue.tail) {
+        return ERROR;
+    }
+    
+    // Get the message
+    msg_t msg;
+    QueueGet(&gRxQueue, &msg);
+
+    // Transfer the message to the provided pointers
+    for(int i = 0; i < 8; i++){
+        data[i] = msg.data[i];
+    }
+    *id = msg.id;
+
+    return SUCCESS;
 }
 
-void CAN1_RX0_IRQHandler(void)
-{
-    CAN_Receive(CAN1, CAN_FIFO0, &RxMessage);
+void CAN1_RX0_IRQHandler(void) {
+    asm("CPSID I");   // Disable Interrupts
+    OSIntEnter();     // Signal to uC/OS
+    asm("CPSIE I");   // Re-enable interrupts
+    
+    // Take any pending messages into a queue
+    while(CAN_MessagePending(CAN1, CAN_FIFO0)) {
+        CAN_Receive(CAN1, CAN_FIFO0, &gRxMessage);
+
+        msg_t rxMsg;
+        rxMsg.id = gRxMessage.StdId;
+        memcpy(&rxMsg.data[0], gRxMessage.Data, 8);
 
-    if ((RxMessage.StdId == 0x001)&&(RxMessage.IDE == CAN_ID_STD) && (RxMessage.DLC == 1)){
-        // TODO: do stuff
-        RxFlag = true;
+        // Place the message in the queue
+        if(QueuePut(&gRxQueue, &rxMsg)) {
+            // If the queue was not already full...
+            // Call the driver-provided function, if it is not null
+            if(gRxEvent != NULL) {
+                gRxEvent();
+            }
+        }
     }
+
+    OSIntExit();      // Signal to uC/OS
+}
+
+void CAN1_TX_IRQHandler(void) {
+    asm("CPSID I");   // Disable Interrupts
```

Comment:
If we try using this on baremetal code will it work? It doesn't seem like there are #defines that will comment out the code used for RTOS.